### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by AdrianMendez1199**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    "server": "node ./dist/index.js",
+    "ts": "ts-node ./src/index.ts",
+    "dev": "NODE_ENV=development nodemon"
+  },
+"dependencies": {
+  "@prisma/client": "^2.0.0-beta.3",
+  "bcrypt": "^4.0.1",
+  "dotenv": "^8.2.0",
+  "graphql-tools": "^5.0.0",
+  "graphql-yoga": "^1.18.3",
```
Comment (wilo087) [inline package.json:21]: Why did you use yoga instead of Apollo?

File: src/graphql/Raffle/Mutation.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+async function createUser(parent: object, args: User, ctx: Context): Promise<User> {
+    const {prisma, request} = ctx
+    const {data} = args
+    
+    isAuth(request)
+
+    try {
+        const userCreated =  await prisma.users.create({
+            data
+        })
```
Comment (wilo087) [inline src/graphql/Raffle/Mutation.ts:19]: Is an standard use tabs on 2 spaces for JS and TS. 

Can't you change that on your linter config please?

Comment (AdrianMendez1199) [reply src/graphql/Raffle/Mutation.ts:19]: ok

File: package.json (hunk 1 above)
Comment (AdrianMendez1199) [reply package.json:21]: just a matter of taste, in the end yoga uses apollo.
yoga is nothing more than a grouping of some tools, such as apollo-server, graphql-suscription among others. I can make the change apollo if you wish.

File: src/graphql/Raffle/Query.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    
+  return prisma.users.findMany({
+    where:{
+      winner: 1
+    }
+  })
+}
+
+
+export const Query = {
```
Comment (Krystian19) [inline src/graphql/Raffle/Query.ts:42]: It's better to just write out the function definition inside of the Query object.

Like so:
```suggestion
export const Query = {
  getUser: (parent, args, ctx): User => { ... }
  // ...
}
```

File: src/utils.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import jwt from 'jsonwebtoken'
+import dotenv from 'dotenv'
+
+dotenv.config()
+
+export const prisma = new PrismaClient()
+const SECRET: string = process.env.SECRET_JWT as string
+
+export interface Request {
+    get(param: string): string;
```
Comment (Krystian19) [inline src/utils.ts:11]: Standart JavaScript indentation is 2 spaces, this looks like 2 tabs.

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+
+const server = new ApolloServer({
+  typeDefs,
+  resolvers,
+  schemaDirectives: {
+    isAuthenticated: IsAuthenticatedDirective
+  },
+  context: (request): object => {
```
Comment (Krystian19) [inline src/index.ts:21]: Always return as early as you can.
```suggestion
  context: (request): object => ({ request, prisma })
```

File: src/utils.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * check if user is auth
+ * @param req 
+ * @param res 
+ * @param next 
+ */
+export function isAuth(token: string): string | object {
+
+  if(!token)
```
Comment (Krystian19) [inline src/utils.ts:60]: Never do body-less conditions.
```suggestion
  if (!token) {
    throw new Error('Unauthorised')
  }
```

File: src/index.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    return {
+      request,
+      prisma
+    }
+  },
+});
+
+server.listen( options.port ).then(({ url }) => {
```
Comment (Krystian19) [inline src/index.ts:29]: Parenthesises should have no spaces around, only curly braces have spaces around.

```suggestion
server.listen(options.port).then(({ url }) => {
```

File: src/utils.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+/**
+ * return jwt token
+ * @param user 
+ */
+export function generateToken ( user: string ): string {
+  return jwt.sign({user}, SECRET, {expiresIn: '24h'})
+}
+
+/**
+ * check if user is auth
```
Comment (Krystian19) [inline src/utils.ts:53]:
```suggestion
 * Checks if user is authenticated.
```

File: src/graphql/Raffle/Mutation.ts [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    
+
+  try {
+    const userCreated =  await prisma.users.create({
+      data
+    })
+
+    return userCreated
+  }catch(e){
+    if(e.code === 'P2002')
```
Comment (Krystian19) [inline src/graphql/Raffle/Mutation.ts:22]: No body-less conditions.

File: src/graphql/Raffle/Mutation.ts [hunk 10]
```diff
@@ -0,0 +1,129 @@
+import { User, Context, generateToken, Auth} from '../../utils'
+import bcrypt from 'bcrypt'
+
+/**
+ * this function create user
+ * @param parent 
+ * @param args 
+ * @param ctx 
+ */
+async function createUser(parent: object, args: User, ctx: Context): Promise<User> {
```
Comment (Krystian19) [inline src/graphql/Raffle/Mutation.ts:10]: The types of the arguments should be deterministic and explicit. Just a plain "Object" is too generic and defeats the whole purpose.

File: src/utils.ts [hunk 11]
```diff
@@ -0,0 +1,67 @@
+import { PrismaClient } from '@prisma/client'
+import jwt from 'jsonwebtoken'
+import dotenv from 'dotenv'
+
+dotenv.config()
+
+export const prisma = new PrismaClient()
+const SECRET: string = process.env.SECRET_JWT as string
```
Comment (Krystian19) [inline src/utils.ts:8]: As far as i know it's already a string. No need to type cast it.

File: src/utils.ts [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+ */
+export function isAuth(token: string): string | object {
+
+  if(!token)
+    throw new Error('Unauthorized')
+
+  token = token.replace('Bearer', '')
+  
+  const data = jwt.verify(token.trim(), SECRET)
+  return data
```
Comment (Krystian19) [inline src/utils.ts:66]: Just return, no need to create a value to just return it:
```suggestion
  return jwt.verify(token.trim(), SECRET)
```

File: src/graphql/Raffle/Mutation.ts [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+  const {prisma} = ctx
+  
+
+  const auth = await prisma.auth.findOne({
+    where:{
+      user
+    }
+  })
+
+  if(!auth || auth == null)
```
Comment (Krystian19) [inline src/graphql/Raffle/Mutation.ts:111]:
```suggestion
  if (!auth)
```

**Pull request merged**